Introduce cloud synchronisation

romcheg opened the pull request:
This patch extends cloud providers to contain necessary settings for cloud syncronisation such as switchable option for enabling or disabling the syncronisation for every specific provider and pluggable syncronisation event processor.  The cloud syncronisation endpoint accepts POST requests for every cloud provider for which the syncronisation is enabled and an event processor is selected. Events are routed for necessary processors in the context of a selected cloud provider.

Commit:
Introduce cloud syncronisation

This patch extends cloud providers to contain necessary settings
for cloud syncronisation such as switchable option for enabling
or disabling the syncronisation for every specific provider and
pluggable syncronisation event processor.

The cloud syncronisation endpoint accepts POST requests for every
cloud provider for which the syncronisation is enabled and an event
processor is selected. Events are routed for necessary processors
in the context of a selected cloud provider.

vi4m commented on src/ralph/virtual/processors/noop.py:
```diff
@@ -0,0 +1,9 @@
+import logging
+
+
+LOGGER = logging.getLogger(__name__)
```
logger = logging.getLogger(__name__)

vi4m commented on src/ralph/settings/base.py:
```diff
@@ -570,3 +570,7 @@ def str_to_bool(s: str) -> bool:
 STATSD_GRAPHS_PREFIX = 'ralph.graphs'
 
 TRANSITION_TEMPLATES = None
+# OpenStack event listeners
```
What is this comment for?

vi4m commented on src/ralph/virtual/subscriptions.py:
```diff
@@ -0,0 +1,40 @@
+
+
+def load_processors():
+    for ep in pkg_resources.iter_entry_points(ENTRY_POINTS_GROUP):
```
s/ep/entry_point

romcheg replied:
For readability purposes I have to leave it as is.

vi4m commented on src/ralph/virtual/subscriptions.py:
```diff
@@ -0,0 +1,40 @@
+    for subscription_config in settings.DCASSET_SYNC_ENDPOINTS:
+        sub_name = subscription_config['name']
+
+        # NOTE(romcheg): Work around the issue when ready() hook is called
```
could you describe a little more what's happengine - or maybe provide issue link for this bug?

vi4m commented on src/ralph/virtual/subscriptions.py:
```diff
@@ -0,0 +1,40 @@
+from django.conf import settings
+
+
+LOGGER = logging.getLogger(__name__)
```
logger

vi4m commented on src/ralph/virtual/processors/noop.py:
```diff
@@ -0,0 +1,9 @@
+
+def endpoint(event_data):
+    """The endpoint for DC asset synchronisation that does nothing."""
+    LOGGER.info('Received new DC asset syncronisation event. Doing nothing.')
```
synchronization

vi4m commented:
Any updates here?

mkurek commented on src/ralph/virtual/models.py:
```diff
@@ -35,11 +35,26 @@
 logger = logging.getLogger(__name__)
 
 
+class CloudSyncProcessor(models.Model, AdminAbsoluteUrlMixin):
```
shouldn't `AdminAbsoluteUrlMixin`be inherited first?

romcheg replied:
Done

mkurek commented on src/ralph/apps.py:
```diff
@@ -3,6 +3,8 @@
 
 from django.apps import AppConfig
 
+from ralph.virtual.subscriptions import generate_listeners
```
wouldn't it be better to move this import into `ready` method and keep imports at the top of the file in `ralph.virtual.subscriptions`?

mkurek commented on src/ralph/virtual/subscriptions.py:
```diff
@@ -0,0 +1,49 @@
+    load_processors()
+
+    from ralph.virtual.models import CloudProvider
+    cloud_providers = CloudProvider.objects.filter(
```
what if I add new cloud provider? why I'll have to restart ralph to create new subscriber?

romcheg replied:
For now yes -- it's necessary to restart Ralph in order to regenerate the list of listeners. The plan is to regenerate them on post_save signal.

romcheg replied:
Right, for now it will be necessary to restart Ralph. I'm going to add a post_save signal subscriber in order to regenerate the list of subscribers every time something is done to a cloud provider. However, I'd rather did it in a separate patch.

mkurek replied:
why couldn't you make one generic endpoint? or pass cloudprovider info as query param?

mkurek commented on src/ralph/virtual/subscriptions.py:
```diff
@@ -0,0 +1,49 @@
+logger = logging.getLogger(__name__)
+
+
+ENTRY_POINTS_GROUP = 'ralph.dchost_processors'
```
so is it dchost processor or cloud processor?

romcheg replied:
It's a cloud processor, I have to rename it.

ar4s commented on src/ralph/virtual/subscriptions.py:
```diff
@@ -0,0 +1,49 @@
+    cloud_providers = CloudProvider.objects.filter(
+        sync_enabled=True,
+        sync_event_processor__isnull=False
+    ).all()
```
please add ``.values_list('name', flat=True)`` to get only name as a flat list

ar4s commented on src/ralph/virtual/subscriptions.py:
```diff
@@ -0,0 +1,49 @@
+    load_processors()
+
+    from ralph.virtual.models import CloudProvider
+    cloud_providers = CloudProvider.objects.filter(
```
``.select_related('sync_event_processor')``

romcheg replied:
Done

romcheg commented:
I have re-done the patch with a different approach -- now there is a custom Django view that routes sync events to appropriate cloud providers basing on URLs. Some tests were also added.

The patch was also rebased and squashed to avoid misunderstanding.

mkurek commented on src/ralph/virtual/processors/noop.py:
```diff
@@ -0,0 +1,9 @@
+
+def endpoint(cloud_provider, event_data):
+    """The endpoint for DC asset synchronisation that does nothing."""
+    logger.info('Received new DC asset syncronisation event. Doing nothing.')
```
synchronisation

romcheg replied:
Done

mkurek commented on src/ralph/virtual/cloudsync.py:
```diff
@@ -0,0 +1,75 @@
+@csrf_exempt
+@require_POST
+def cloud_sync_router(request, cloud_provider_id):
+    load_processors()
```
can't we load it in `AppConfig.ready()` or sth like that?

romcheg replied:
Using AppConfig.ready() is bound with a few issues. One of them is that it's not even possible to import any object on a module level. In order to avoid that I have specifically re-written the patch to not use ready(). Loading the list of entry points on the first request won't hurt at all because it's won't bring any notable performance problems.

mkurek commented on src/ralph/virtual/cloudsync.py:
```diff
@@ -0,0 +1,75 @@
+    except CloudProvider.DoesNotExist:
+        return HttpResponseNotFound()
+
+    CLOUD_SYNC_PROCESSORS[cloud_provider.sync_event_processor.module](
```
what if `sync_event_processor` is not set?

romcheg replied:
Good point. I will add an extra DB filter for that.

mkurek commented on src/ralph/virtual/models.py:
```diff
@@ -35,11 +35,26 @@
 logger = logging.getLogger(__name__)
 
 
+class CloudSyncProcessor(AdminAbsoluteUrlMixin, models.Model):
```
what's the point of this model? Couldn't we point processor by name in the url?

romcheg replied:
I wanter users to have a convenient admin pages where they can select processors from the list of available ones. On the other hand binding the field directly to the list of available entry points would require an automatic synchronisation of the data in the DB with the list of entry points.

It's a good point to extend the router with a check whether or not the specified processor is actually available.

mkurek replied:
But if processor is pointed in the url, user will "select" it in hermes console (by entering url). Right now you still have to sync data between db and code (ex. what if I change the name of the processor or remove one)? The compromise here would be storying `sync_event_processor` as string in `CloudProvider` model :)

MarekBleschke reviewed (dismissed):
looks ok

mkurek commented on src/ralph/virtual/cloudsync.py:
```diff
@@ -0,0 +1,78 @@
+    if CLOUD_SYNC_PROCESSORS is not None:
+        return
+
+    with LOCK:
```
why do you need this?

mkurek commented on src/ralph/virtual/cloudsync.py:
```diff
@@ -0,0 +1,78 @@
+
+        logger.info('Loading cloud sync processors.')
+
+        CLOUD_SYNC_PROCESSORS = {}
```
to avoid race condition, couldn't you use here some local variable, and at the end of this function, assign value of this local variable to `CLOUD_SYNC_PROCESSORS`?

mkurek commented on src/ralph/virtual/cloudsync.py:
```diff
@@ -0,0 +1,78 @@
+    except CloudProvider.DoesNotExist:
+        return HttpResponseNotFound()
+    except KeyError:
+        return HttpResponse('Specified processor is not available', status=501)
```
doesn't 400 or 404 better fits here?

mkurek commented on src/ralph/virtual/processors/noop.py:
```diff
@@ -0,0 +1,9 @@
+logger = logging.getLogger(__name__)
+
+
+def endpoint(cloud_provider, event_data):
```
could you provide better name for this processor?

pnuckowski commented on src/ralph/virtual/cloudsync.py:
```diff
@@ -0,0 +1,78 @@
+
+
+ENTRY_POINTS_GROUP = 'ralph.cloud_sync_processors'
+CLOUD_SYNC_PROCESSORS = None
```
To drop that global variable maybe use `nonlocal` ?
```python
def processor_factory():
  processors = None
  def inner():
    nonlocal processors
    if processors is None:
      processors = load_processors()
    return processors

get_processors = processor_factory()


# then in cloud_sync_router method
processors = get_processors()
```
